Allow async functions to throw

cpcallen opened the pull request:
Provide a convenient means to allow asynchronous functions to throw an exception rather than return a value.

cpcallen committed:
Allow async functions to throw

Provide a convenient means to allow asynchronous functions to throw an exception rather than return a value.

clvs7-gh commented on interpreter.js:
```diff
@@ -3226,8 +3226,13 @@ Interpreter.prototype['stepCallExpression'] = function(stack, state, node) {
-      var callback = function(value) {
-        state.value = value;
+      var callback = function(value, exception) {
+        if (arguments.length === 1) {
```
Hello.

I think this condition makes (maybe unintended) breaking change for **void** async functions.
For example:

```
const { Interpreter } = require('./interpreter');
global.acorn = require('./acorn');

const myCode = `
// something to do
sleep(1000);
// something to do
`;

const initFunc = (i, s) => {
  i.setProperty(s, 'sleep', i.createAsyncFunction((timeMs, callback) => setTimeout(callback, timeMs)));
};
const myInterpreter = new Interpreter(myCode, initFunc);
myInterpreter.run();
```

Above code works well on current JS-Interpreter, but it won't work after apply this PR.

Of course changing `setTimeout(callback, timeMs)` to `setTimeout(() => callback(undefined), timeMs)` works, but presumably it's not good.

cpcallen replied:
Strictly speaking there's no such thing as a void function in JavaScript (every function has a return value, even if you don't specify one intentionally, but I take your point about usage compatibility, which is an objective for this PR.  Fortunately it should be easy to fix.

cpcallen replied:
Fixed (I believe!)

Thanks for your feedback.

cpcallen committed:
Async callback with no args should return

Existing usage includes calling the callback with no arguments, intending it to return undefined.  Preserve this behaviour by treating zero and one arguments equivalently, and only throwing if it is called with two or more.

(This will break any existing usage which calls the callback with more than one argument, but that ought to be fairly uncommon!)

Webifi commented:
@cpcallen 
Trying to figure out how to implement something like this in the changes I have in #201.    Problem is, #201 allows for multiple async states to be waiting for callbacks.  Unwind becomes problematic when using this async exception handling, since it assumes there's only one async state waiting.

Any thoughts?
